### ISSUE TITLE
Missing coercion or validation for default values

### ISSUE THREAD

**Issue by eapache** [rmosolgo/graphql-ruby#3248]
**Describe the bug**

@benjie pointed out in https://github.com/graphql/graphql-spec/pull/793 that the GraphQL Spec and JS reference implementations are both somewhat confusing/broken when dealing with mistyped default values. I haven't gone through and tried to reproduce all the various combinations of this issue, but I did demonstrate a simple case to show that graphql-ruby suffers from at least a similar class of bug.

**Versions**

`graphql` version: 1.11.4 w/ interpreter, but I'm pretty sure it applies to master too

**GraphQL schema**

```ruby
  field :test, :integer, null: true do |field|
    field.argument(:arg, :integer, required: true, default_value: "string")
  end

  def _test(arg:)
    arg
  end
```

**GraphQL query**

```graphql
query {
  test
}
```

```json
{
  "data": {
    "test": null
  }
}
```

**Expected vs actual behaviour**

There are two interesting points in this example:
- GraphiQL actually says the sample query is invalid; apparently the `default_value` is not making it through into the introspection output, so GraphiQL claims the argument needs a value. However, if you actually *run* it then it works (sorta).
- When you run it, the `default_value` doesn't make it through into the resolver method, which gets `nil` instead. This could really mess with resolver code which assumes a value (since the argument is required-with-a-default at the GraphQL layer).

---

@rmosolgo there is still some debate on what the final specified behaviour should be, but I wanted to put this on your radar. The fact that the default value is getting stripped out in places suggests to me that we are validating it already somewhere, we're just not handling validation failures aggressively enough?

cc @swalkinshaw.

**Comment by rmosolgo** [comment]
Thanks for the heads up on this! I bet the default value gets dropped in `coerce_input`, which returns `nil` for unmanageable values: 

https://github.com/rmosolgo/graphql-ruby/blob/74bbeea30db3cd90c55e8008d55fa14528479802/lib/graphql/types/int.rb#L13

Like you said, I bet it's never validated.

Is there any reason _not_ to apply validation during configuration (ie, during the `argument(...)` method evaluation)?

**Comment by eapache** [comment]
Two potential complications came up during the Working Group meeting:
- Default values are mostly specified as language-native types, not as GraphQL AST tokens or on-the-wire transport bytes, so what it means to validate them is confusing/different from the normally-specified flow. (Though this is maybe more of a spec issue than an implementation issue.)
- You can do some messy things with default types that themselves have default fields which can lead to potentially infinite recursion in the validator if we're not careful. Something like
```ruby
class A < GraphApi::InputObject
  field :a, A, required: true, default_value: A.new
end
```

**Comment by jeffsirocki** [comment]
Hey @eapache and @rmosolgo,
I'm wondering if there should be an additional note in https://github.com/rmosolgo/graphql-ruby/blob/master/guides/queries/interpreter.md pointing out that when migrating to the new Interpreter some default_value configurations may need to be updated - maybe something like?
```
Some default_value configurations must be updated too. For example:

- argument :field, String, default_value: :id, required: false
+ argument :field, String, default_value: 'id', required: false

(If the argument has a default value, {defaultValue} must be compatible with {argumentType} as per the coercion rules for that type, and coercion of {defaultValue} must not cause an infinite loop.)
```
When opting-in to use the Interpreter (with `graphql` v1.11.6) I also came across this in the following line: 
```argument :sort_by_column, String, default_value: :id, required: false``` - in my opinion adding this in the docs may be useful to others migrating.

**Comment by rmosolgo** [comment]
It sounds like a specific behavior is in draft in the spec: https://github.com/graphql/graphql-wg/issues/645#issuecomment-815813397

**Comment by benjie** [comment]
> Thanks for the heads up on this! I bet the default value gets dropped in coerce_input, which returns nil for unmanageable values:

The GraphQL Spec states that this situation should throw, I think:

https://spec.graphql.org/draft/#sel-NANRHHCJFTDFBBCAACEW0yS

**Comment by eapache** [comment]
@rmosolgo I took a look into fixing this and ran into a bunch of interesting points:
- You'd think in most schemas we should be able to validate default values at schema definition time, but all the coercion methods that actually do this depend on having a context available. Late-bound types also make this impossible.
- Runtime argument objects do no validation at all and are not set up for error handling, because they assume the static query validation has taken care of it.
- Static query validation only has access to the arguments defined in the query, which precisely excludes the arguments whose default values we care about.

Options I've considered:
- Set up a dummy context and just do it at schema-build time (or whenever the late-bound argument type is resolved, if I can find the right hook). This is mostly the best, except that it will break for any weird custom scalars which have context-dependent coercion logic.
- Write a new static validation rule which looks at every argument-containing element in the query and fully recurses its arguments and validates the default one... gross.
- Plumb through error handling code deep into runtime argument objects just for this one case... gross.

Is there another option I'm missing? Do we care about context in custom scalar coercion (at a quick glance none of Shopify's scalars use it)?

**Comment by rmosolgo** [comment]
I agree that your first suggestion is the best. A dummy context will likely be good enough, and if those intense cases come up, we can investigate one of those more intense approaches. 

Here's a dummy context used elsewhere for purposes like this: 

https://github.com/rmosolgo/graphql-ruby/blob/bace1e4027900fc8779a5c2fd393ff6456046cea/lib/graphql/query/null_context.rb#L5

**Comment by eapache** [comment]
👍 can you give me a hint where I should hook to handle late-bound types?

**Comment by eapache** [comment]
Possible a simple question, possibly a complicated one: should default values be `prepare`d?

**Comment by eapache** [comment]
@rmosolgo I spoke with @leebyron and a few other folks working on implementations of this at the Working Group meeting yesterday. graphql-js and most others have moved to a two-pass schema building system (resolve all the types, then do a complete second pass for validations and such). This makes a lot of sense and would provide us with a cleaner solution, but is a _much_ larger refactor than I can contribute right now.

It sounds like you're OK with https://github.com/rmosolgo/graphql-ruby/pull/3476 or perhaps a variation like https://github.com/rmosolgo/graphql-ruby/commit/30722fbd571826fc037909a8f072b104045001d6 - I'm happy to polish up and add tests for whatever set of trade-offs you'd prefer to accept. Just let me know what direction you want to go.
